src/PostgREST/Query/QueryBuilder.hs

```diff
@@ -210,10 +211,17 @@ callPlanToQuery (FunctionCall qi params arguments returnsScalar returnsSetOfScal
         -- We could fallback to providing this NULL value in those cases.
         encodeArg Nothing = "NULL"
 
+    -- the columns here would be the returnings + the columns that would later
+    -- be used by a where clause filter, if they intersect, we remove the duplicates
+    -- and if * is returned then no need to explicitly add filter columns
     returnedColumns :: SQL.Snippet
-    returnedColumns
-      | null returnings = "*"
-      | otherwise       = intercalateSnippet ", " (pgFmtColumn (QualifiedIdentifier mempty "pgrst_call") <$> returnings)
+    returnedColumns = case returnings of
+      []    -> "*"
+      ["*"] -> pgFmtColumn (QualifiedIdentifier mempty "pgrst_call") "*"
+      _     -> intercalateSnippet ", " (pgFmtColumn (QualifiedIdentifier mempty "pgrst_call") <$> returnedColumns')
+        where
+          returnedColumns' = removeDups $ returnings ++ filterFields
+          removeDups = L.nub -- function to remove duplicates from list (possibly impact performance for very large lists)
```

Comment:
Since these fields are meant to be unique. Couldn't we use `Data.Set` instead to avoid this deduplication?

So `funCReturning` (returnings here) could be a Set instead of a List, then here we would just do an union with `S.fromList filterFields`?

That might be cleaner.

Additionally, not sure if large lists are a concern here since we don't expect so much fields thanks to the URL length limit.

Reply:
Yes, I think it would be cleaner to this in a refactor first to change the `returnings` to a `Set`.